### PR TITLE
F 22

### PR DESCRIPTION
**Pull request by jaypif**

### CODE REVIEW

File: opennebula/resource_opennebula_image.go [hunk 1]
```diff
@@ -558,6 +609,19 @@ func resourceOpennebulaImageUpdate(d *schema.ResourceData, meta interface{}) err
 		log.Printf("[INFO] Successfully updated Image Type %s\n", image.Name)
 	}
 
+	if d.HasChange("tags") {
+		tagsInterface := d.Get("tags").(map[string]interface{})
+		for k, v := range tagsInterface {
+			// It automatically erase the value of the previous one
+			image.Template.AddPair(strings.ToUpper(k), v)
```
Comment (treywelsh) [inline opennebula/resource_opennebula_image.go:599]: it seems that it miss a image.Template.Del before this line ?

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,11 @@ FEATURES:
 OpenNebula provider issue: ([#16](https://github.com/terraform-providers/terraform-provider-opennebula/issues/16))
 * resource/opennebula_virtual_machine: Associate a VM group (only during VM creation) ([#16](https://github.com/terraform-providers/terraform-provider-opennebula/issues/16))
 * resource/opennebula_template: Associate a VM group.
+* resource/opennebula_image: Add support for tags ([#22](https://github.com/terraform-providers/terraform-provider-opennebula/issues/22))
+* resource/opennebula_security_group: Add support for tags ([#22](https://github.com/terraform-providers/terraform-provider-opennebula/issues/22))
+* resource/opennebula_template: Add support for tags ([#22](https://github.com/terraform-providers/terraform-provider-opennebula/issues/22))
+* resource/opennebula_virtual_machine: Add support for tags ([#22](https://github.com/terraform-providers/terraform-provider-opennebula/issues/22))
+* resource/opennebula_virtual_machine_group: Add support for tags ([#22](https://github.com/terraform-providers/terraform-provider-opennebula/issues/22))
```
Comment (treywelsh) [inline CHANGELOG.md:13]: VNet not mentionned

File: opennebula/resource_opennebula_virtual_machine_test.go [hunk 3]
```diff
@@ -88,7 +95,8 @@ func TestAccVirtualMachinePending(t *testing.T) {
 		CheckDestroy: testAccCheckVirtualMachineDestroy,
 		Steps: []resource.TestStep{
 			{
-				Config: testAccVirtualMachinePending,
+				Config:             testAccVirtualMachinePending,
+				ExpectNonEmptyPlan: true,
```
Comment (treywelsh) [inline opennebula/resource_opennebula_virtual_machine_test.go:99]: For my own understanding: why adding this `ExpectNonEmptyPlan` here ?

File: opennebula/resource_opennebula_image.go [hunk 4]
```diff
@@ -451,6 +453,55 @@ func resourceOpennebulaImageRead(d *schema.ResourceData, meta interface{}) error
...
+		}
+	}
+
+	if len(tags) > 0 {
+		err := d.Set("tags", tags)
+		if err != nil {
+			return err
+		}
+	}
+
```
Comment (treywelsh) [inline opennebula/resource_opennebula_image.go:504]: here, the way to process tags differ from other resources:
there is no method `flattenXXX` called at end of the function that handle all remaining stuffs in the template.

why ?

File: opennebula/resource_opennebula_image.go [hunk 5]
```diff
@@ -635,6 +699,11 @@ func generateImageTemplate(d *schema.ResourceData) (string, error) {
 		tpl.Add(imk.Target, val.(string))
 	}
 
+	tagsInterface := d.Get("tags").(map[string]interface{})
+	for k, v := range tagsInterface {
+		tpl.AddPair(strings.ToUpper(k), v)
```
Comment (treywelsh) [inline opennebula/resource_opennebula_image.go:687]: AddPair already use ToUpper function internally.
But, Del, and other GetPair etc... functions does not.

We may need to fix this in Goca later, as the behaviour is not consistent for all methods...

File: opennebula/resource_opennebula_image.go (hunk 1 above)
Comment (jaypif) [reply opennebula/resource_opennebula_image.go:599]: it is not required but you are right. I will add it to be aligned with other resource. Thanks

File: CHANGELOG.md (hunk 2 above)
Comment (jaypif) [reply CHANGELOG.md:13]: indeed

File: opennebula/resource_opennebula_image.go (hunk 5 above)
Comment (jaypif) [reply opennebula/resource_opennebula_image.go:687]: yes. But it not  linked directly to this patch ;)

File: opennebula/resource_opennebula_virtual_machine.go [hunk 6]
```diff
@@ -314,6 +315,41 @@ func resourceOpennebulaVirtualMachineRead(d *schema.ResourceData, meta interface
 		return err
 	}
 	return nil
+
```
Comment (treywelsh) [inline opennebula/resource_opennebula_virtual_machine.go:317]: this is dead code due to the return nil above

Comment (jaypif) [reply opennebula/resource_opennebula_virtual_machine.go:317]: you are totally right

File: opennebula/resource_opennebula_security_group.go [hunk 7]
```diff
@@ -257,13 +258,42 @@ func resourceOpennebulaSecurityGroupRead(d *schema.ResourceData, meta interface{
...
 
-	description, _ := securitygroup.Template.GetStr("DESCRITPION")
+	description, _ := securitygroup.Template.Get(sgk.Description)
 	d.Set("description", description)
 
 	if err := d.Set("rule", generateSecurityGroupMapFromStructs(securitygroup.Template.GetRules())); err != nil {
 		log.Printf("[WARN] Error setting rule for Security Group %x, error: %s", securitygroup.ID, err)
 	}
 
+	err = flattenSecurityGroupTags(d, &securitygroup.Template)
```
Comment (treywelsh) [inline opennebula/resource_opennebula_security_group.go:268]: DESCRIPTION wiil be in tags as it's not deleted

Comment (jaypif) [reply opennebula/resource_opennebula_security_group.go:268]: Actually not.
Only declared tags in the TF file are searched in the flatten function

**Review by treywelsh — APPROVED** [review]

**Pull request merged**